### ISSUE TITLE
Hi, this crashes for me

### ISSUE THREAD

**Issue by ilanle** [primaryobjects/lda#2]
Hello and thank you for creating this package.
I am trying to run lda for a bulk of textual documents and I get crashes of this sort :+1: 

docs= [array of text documents]
ldaResult=lda(docs,10,50)

this.nw[this.documents[m][n]][topic]++;
TypeError: Cannot read property '2' of undefined
    at [object Object].initialState (C:\web\node_modules\lda\lib\lda.js:143:46)
    at [object Object].gibbs (C:\web\node_modules\lda\lib\lda.js:161:14)

Do you have an idea why this happens?

**Comment by ilanle** [comment]
If I dig in a bit fu

**Comment by ilanle** [comment]
This happens after some iteration, one of the values in this.documents[m][n] equals -1 which kills it.

I have modified the code a bit to have printouts:
            for (var n = 0; n < N; n++) {
                var topic = parseInt(""+(Math.random() \* K));  
                this.z[m][n] = topic;
                console.log('current topic is '+topic);
                console.log('m is '+m+' and n is '+n);
                console.log('current cell index is '+this.documents[m][n]);
                console.log('current cell is '+this.nw[this.documents[m][n]]);
                this.nw[this.documents[m][n]][topic]++;
                this.nd[m][topic]++;
                this.nwsum[topic]++;
            }

And this is the output at some iteration:

current topic is 6
m is 53 and n is 146
current cell index is -1
current cell is undefined

C:\web\node_modules\lda\lib\lda.js:146
                this.nw[this.documents[m][n]][topic]++;
                                             ^
TypeError: Cannot read property '6' of undefined

**Comment by primaryobjects** [comment]
This could be due to a piece of text ending up with a null word. Possibly related to https://github.com/primaryobjects/lda/pull/1? Could you check your data set for any sentences with special characters, that you might be able to clean-up before running through lda?

**Comment by ilanle** [comment]
Thanks for the quick reply.
I thought that might be the case, let me have a look.
Just to make sure, the example given for this library uses sentences out of a single document, but of course I mean to use full documents out of a large corpus of documents.
This by itself shouldn't pose a problem right? I am asking this because I see that the code itself has reference inside it to sentences

[five minutes later...]
So I think I found a single record which crashed it, by elimination (saw in which line of the record set it crashed and removed that line)
Right now I am running it and it doesn't crash , but seems suspended (with moderate CPU usage).
Do I expect it to run for a significant time?
How can I provide myself with indication as to progress?
Thanks !!

**Comment by primaryobjects** [comment]
You may want to try a smaller subset of your data, just to get some quicker results. Yes, it can take a while to process a large dataset. You could always try adding some console.log() statements in the larger for loops to indicate progress. You could try adding a log statement [here](https://github.com/primaryobjects/lda/blob/master/lib/lda.js#L24) and [here](https://github.com/primaryobjects/lda/blob/master/lib/lda.js#L164) for example.

Btw, could you list the sentence that caused the error? Maybe I can add a fix to ignore that text or if you want to make a pull-request with a fix.

**Comment by ilanle** [comment]
I will try to find that line again and post it.
I have another question , sorry for taking advantage of your attentiveness :-)
So I guess there is an important connection between the number of topics, number of terms, iterations and my data.
My data is taken from Wikipedia articles, so each entry is about 100 word long and the words are quite spread.
I am trying different configurations and always end up with terms which have probability NaN/null.
Are there guidelines for picking out the numbers?
Thanks

EDIT:

I found the offending record, didn't find any irregular characters in it, and started narrowing it down (by division), and ended up with (please sit down for this):  
When any record includes the word "constructor", the process breaks.
Whaaat??

**Comment by ilanle** [comment]
This seems to be something more fundamental in javascript 
dictionary['constructor']

Ha Ha

So I guess you have to all of those strings related to the object nature of dictionary to your stopwords or something

**Comment by primaryobjects** [comment]
Regarding the first issue about NaN/null, I ran a simple lda program against a random Wikipedia page and it seemed to work correctly. Here's the [example](https://gist.github.com/primaryobjects/2d4bcdcae0b0316bc642).

As far as the 'constructor', you are correct! That is so weird. Maybe it should just ignore that word.

Update:

I just edited stopwords.js and [added](https://github.com/primaryobjects/lda/commit/30ba46fcd9d27d46904d255d42584266e28de4f6) 'constructor' as a stop-word. This appears to fix the problem.

**Comment by ilanle** [comment]
Amazing !

So regarding NaN, 
I tried the following settings for example  : 
ldaResult=lda(docs,20,100)
but I did also play with the number of iterations within the code itself, set it to 1000 instead of 10000.
Could this be it?

Here is a typical record I use:

k 12 st agnes k 8 st dominic st agnes academy st dominic school private roman catholic school consisting girls high school co educational elementary middle school memphis tennessee united states school located roman catholic diocese memphis follows catholic principles run diocese st agnes academy st dominic school serves girls kindergarten twelfth grade boys kindergarten eighth grade st agnes academy founded dominican sisters january 1851 chartered january 1852 founding sisters magdalen clark catherine mccormack vincentia fitzpatrick ann simpson sr lucy harper emily thorpe sisters magdalen catherine professed st mary springs four st catharine kentucky dominican sisters st mary springs st catharine kentucky eventually merged along six congregations become dominican sisters peace congregation continues sponsor school today sr catherine died august mother angela lynch replaced sr frances conlon st agnes situated suburbs memphis vance orleans mile quarter court square center city doors opened february 4 1851 20 boarders 15 day students school year ended july 7 1851 enrollment increased 50 addition school completed dominican sisters also boarded orphans 1864 sisters opened st peter s orphanage fall 1867 fall 1873 memphis enveloped yellow fever epidemic times many dominican sisters died rendering service sick memphis may 16 1878 yellow fever faded academy caught fire reduced ashes another fire 1900 caused less damage ne

**Comment by primaryobjects** [comment]
I think the problem is the settings 20,100. That means to detect 20 topics with 100 terms each. The record you displayed above doesn't have a whole lot of text, so the algorithm might not be able to find enough terms to group into that many topics. Also, the text seems broken up, maybe you've already pre-processed it with something.

Have you tried using a setting like lda(docs, 2, 5) ?

**Comment by ilanle** [comment]
I did try that, still getting NaNs.
Could it have to do with the fact that I am trying less iterations?
I have to leave my desk now, but I thank you for your help, I will try to figure this out later and give you an update

**Comment by ilanle** [comment]
OK,
I was able to get results now, 
I will continue tweaking the numbers and playing. Thank you for your help.

BTW, the bug was even more bizzare, because the actual word in my text which broke the algo wasn't "constructor" but rather "constructors". I guess your stemmer changed it into "constructor", so please make sure that the stopwords are filtered AFTER the stemmer.

Thanks again,
Ilan

**Comment by ilanle** [comment]
Kory Hi again,
Just a heads-up.
I tested with the word "constructors" (plural) after pulling the latest revision and it still crashes.
I believe the reason for that is that you have the stemmer running after the stopwords filter.
So you'll probably need to either add all of the forms which stem to "constructor" or move the stopwords filter to run after the stemmer.

I don't know if you see comments after closing, so I reopened. 
Cheers,
Ilan

**Comment by primaryobjects** [comment]
Thanks, you're right!

I took a peek at the code and realized it was, indeed, only checking the un-stemmed word against the stop-words. I pushed a [fix](https://github.com/primaryobjects/lda/commit/7b442c34130f2dc80532d20b24a59072784c95ca) that now checks both the original and stemmed term against the stop-words list. I tested and it's no longer crashing on any variation of 'constructor'. Thanks.
